virttest/env_process.py

```diff
@@ -569,6 +571,12 @@ def preprocess(test, params, env):
     """
     error_context.context("preprocessing")
 
+    global _setuper_mgr
+    _setuper_mgr = test_setup.SetuperMgr(test, params, env)
+    if params.get("setup_ksm") == "yes":
+        _setuper_mgr.register(test_setup.KSMSetuper)
+    _setuper_mgr.do_setup()
```

Comment:
it looks this line should be in if block

Reply:
The `do_setup` method will execute all the setup procedures of each setuper so it should be called only once unconditionally, the code structure should be:
```python
if params.get("setup_s1") == "yes":
    _setuper_mgr.register(S1)
if "setup_s2" in params:
    _setuper_mgr.register(S2)
_setuper_mgr.register(S3)

...

# Do setup
_setuper_mgr.do_setup()
```